### PR TITLE
[Gregorian Calendar] WIP: Implement `ordinality(of:in:for:)`

### PR DESCRIPTION
**Pull request by itingliu**
### 1st: Drive-by fix for quarter handling in `dateComponents(from:)`
The current handling for quarter is wrong for leap years. Fix this by reversing the leap year check.

### 2nd: WIP: Implement `ordinality(of:in:for:)`
 The implementation follows that in Calendar_ICU. But instead of mutating the stored ICU calendar, we pass around a `Date` and its `DateComponents` for keeping track of the calculation state.
    
Unfinished tasks
- There are a few places where we try to find the next date matching a given date component repeatedly in while loops. We need to add failsafe to prevent infinite loop.

### CODE REVIEW

File: Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift [hunk 1]
```diff
@@ -319,10 +335,769 @@ internal final class _CalendarGregorian: _CalendarProtocol, @unchecked Sendable
         fatalError()
     }
     
+    func minMaxRange(of component: Calendar.Component, in dateComponent: DateComponents) -> Range<Int>? {
```
Comment (Harry-KNIGHT) [inline Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift:338]:
```suggestion
    func minMaxRange(
                                     of component: Calendar.Component,
                                     in dateComponent: DateComponents
                                     ) -> Range<Int>? {
```

File: Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift [hunk 2]
```diff
@@ -319,10 +335,769 @@ internal final class _CalendarGregorian: _CalendarProtocol, @unchecked Sendable
...
+                return nil
+            }
+
+        case .yearForWeekOfYear:
+            switch smaller {
+            case .quarter:
+                return nil
+            case .month:
+                return nil
+            case .weekOfMonth: /* kCFCalendarUnitWeek_Deprecated */
```
Comment (Harry-KNIGHT) [inline Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift:813]:
```suggestion
            case .weekOfMonth:
```

File: Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (parkera) [reply Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift:338]: Actually I don't think we should adopt this style - it's not idiomatic across the rest of our sources, for better or for worse.

**Review by parkera — APPROVED** [review]

**Review by parkera — APPROVED** [review]

File: Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (itingliu) [reply Sources/FoundationEssentials/Calendar/Calendar_Gregorian.swift:287]: I agree with @parkera

**Comment by itingliu** [comment]
@swift-ci please test

**Pull request merged**